draft-specifications/CapTP Specification.md

```diff
@@ -0,0 +1,1002 @@
+```
+
+### Sending
+#### `promise-pos`
+This should be the `desc:answer` or a`desc:import-promise` value which is the
+promise you wish to pick the value from.
+#### `selected-value-pos`
+This should be a zero indexed integer which specifies which value should be
+picked out of the multiple values.
+#### `new-answer-pos`
+This should be a new unique answer position that the selected value should be
+exported at.
+
+### Receiving
+When the `op:pick` message is received, a promise should be exported at the
+answer position specified by `new-answer-pos`. The promise should eventually
+resolve to the value at the index specified by `selected-value-pos`, in values
+provided in the `promise-pos` promise. If the `promise-pos` promise breaks, or
+the `selected-value-pos` is out of bounds, the promise should break.
+
+## [`op:abort`](#op-abort)
+
+This is used to abort a CapTP session, when this is sent the connection should
+be severed and any per session information (e.g. session key pair, etc.) should
+be removed.
+
+The `op:abort` message is:
+
+```text
+<op:abort reason>  ; reason: String
+```
+
+## [`op:listen`](#op-listen)
+
+This is used to listen to a promise. This is done in order to get notified when
+the promise is fulfilled with a value or broken on an error. Please see the
+[promises section](#promises) for more information on how promises work.
+
+The `op:listen` message is:
+
+```text
+<op:listen to-desc           ; desc:export | desc:answer
+           listen-desc       ; desc:import-object
+           wants-partial?    ; boolean
+```
+
+The `wants-partial?` flag indicates if a "partial" update should be provided as
+the notification when the promise resolves. If `wants-partial?` is false and the
+promise resolves to another promise, a notification to the `listen-desc` is not
+sent. A notification is only ever sent to the `listen-desc` when the promise
+resolves to a non-promise value, or breaks. If `wants-partial?` is true, a
+notification on any resolution, including a promise. In the case it does notify
+on a promise the listen is considered complete, if future notifications are
+desired, further `op:listen` operations should be sent to the promises.
+
+### Sending
+
+`to-desc` MUST be a `desc:export` or `desc:export` which corresponds to a
+promise on the remote side.
+
+`listen-desc` MUST be a `desc:import` that is being imported. This will be
+invoked when the promise comes to a resolution.
+
+### Receiving
+
+When receiving this message, providing `to-desc` matches a promise exported to
+this session, a mechanism MUST be setup to fulfill or break the provided
+resolver when a resolution is available to the `listen-desc` object. If a
+resolution is already available, the resolver provided in `listen-desc` MUST be
+fulfilled or broken.
+
+## [`op:gc-export`](#op-gc-export)
+
+When a reference is given out to another CapTP session, the reference must be
+kept valid until the other session is done with it. This is achieved by
```

Comment:
"other session", or other party within a session?

Reply:
yeah I agree, I've used session in a number of places in ambiguous ways. To me a session consists of two parties speaking CapTP with one another, however I've used it to mean a single party within a session which is confusing. I've fixed this wording here.